src/compute-client/src/controller/instance.rs

```diff
@@ -1619,9 +1619,18 @@ where
             }
             None => {
                 // Write-only collections cannot be read within the context of the compute
-                // controller, so we can immediately advance their read frontier to the new write
-                // frontier.
-                new_frontier.clone()
+                // controller, so their read frontier only controls the read holds taken on their
+                // inputs. We can safely downgrade the input read holds to any time less than the
+                // write frontier.
+                //
+                // Note that some write-only collections (continual tasks) need to observe changes
+                // at their current write frontier during hydration. Thus, we cannot downgrade the
+                // read frontier to the write frontier and instead step it back by one.
+                Antichain::from_iter(
+                    new_frontier
+                        .iter()
+                        .map(|t| t.step_back().unwrap_or(T::minimum())),
+                )
```

Comment:
Indeed, I wanted to avoid having special cases apart from the one we already have in the controller (indexes vs "write-only collections"). If it turns out that we need to treat CTs differently from other write-only collections, we'll start doing that, but so far I don't see that need.